### ISSUE TITLE
Suggestion: Get rid of the save button?

### ISSUE THREAD

**Issue by jefferyharrell** [RimSort/RimSort#160]
Strictly speaking, is there a reason RimSort has to have a "Save mod list" button? I've gotten used to programs saving work automatically. I think maybe RimSort would be better off implementing an undo feature than requiring the user to hit "Save mod list" after making changes.

**Comment by loveboylion** [comment]
1) This would trigger lot of disk activity and can also lead to corrupt mod list since , any addition or removal of mod would trigger a save action , rearranging mods or sorting will also trigger save action

2) also how would undo feature work u create auto backup on load or after each change resulting in multiple temp mod list files which would lead to unwanted disk activity as well as of storage

basically any change would trigger a save action and same with the undo feature

**Comment by jefferyharrell** [comment]
Yes, I'm thinking that changing the active mods list (adding a mod, removing a mod, manually changing the order of items in the list or clicking the "sort" button) would trigger an autosave that would write a temp file, rename `ModsConfig.xml` to `ModsConfig_backup.xml` or whatever, then move the temp file to `ModsConfig.xml`. I think the amount of disk activity would be trivial; `ModsConfig.xml` is on the order of kilobytes, tops, even for the biggest mod lists.

As for your second point, undo states don't need to be cached to disk. They can be stored in memory. One brute-force hit-it-with-a-rock option would be to store lists of active mods on a stack; every time the user modifies the active mods list, push the previous list onto the stack. `Undo` pops the stack. There are more clever ways to do it, obviously, involving saving the equivalent of a transaction log, but again we're talking about _bytes_ of data per item on the stack, right?

This is good, though. Let's keep the conversation going.

**Comment by twstagg** [comment]
> 1. This would trigger lot of disk activity and can also lead to corrupt mod list since , any addition or removal of mod would trigger a save action , rearranging mods or sorting will also trigger save action
> 
>     2. also how would undo feature work u create auto backup on load or after each change resulting in multiple temp mod list files which would lead to unwanted disk activity as well as of storage
> 
> 
> basically any change would trigger a save action and same with the undo feature

1. The amount of disk activity would be negligible. It is just writing to a single, very small file (unless you try and load a crazy amount of mods somehow. To  clean up any clutter from however many "states" this would support, we could save the states in a subfolder right next to `ModsConfig.xml` called `.ModsConfigStates` or something like that that would be supported by RimSort.
2. As what @jefferyharrell has said, we could cache the states in memory somehow. We are already doing so with the one that is saved when you load up the program (See the "Restore active state" option in the app)
Additional reference for 2.
https://github.com/RimSort/RimSort/blob/main/view/main_content_panel.py#L772-L775

**Comment by twstagg** [comment]
> Strictly speaking, is there a reason RimSort has to have a "Save mod list" button? I've gotten used to programs saving work automatically. I think maybe RimSort would be better off implementing an undo feature than requiring the user to hit "Save mod list" after making changes.



> Yes, I'm thinking that changing the active mods list (adding a mod, removing a mod, manually changing the order of items in the list or clicking the "sort" button) would trigger an autosave that would write a temp file, rename `ModsConfig.xml` to `ModsConfig_backup.xml` or whatever, then move the temp file to `ModsConfig.xml`. I think the amount of disk activity would be trivial; `ModsConfig.xml` is on the order of kilobytes, tops, even for the biggest mod lists.
> 
> As for your second point, undo states don't need to be cached to disk. They can be stored in memory. One brute-force hit-it-with-a-rock option would be to store lists of active mods on a stack; every time the user modifies the active mods list, push the previous list onto the stack. `Undo` pops the stack. There are more clever ways to do it, obviously, involving saving the equivalent of a transaction log, but again we're talking about _bytes_ of data per item on the stack, right?
> 
> This is good, though. Let's keep the conversation going.

I'm all for this being implemented. I think that it should also:

* Keep a configurable amount of "states" to display (with timestamps)
    * This is essentially going to be like a "mod list clipboard"?
* I think it would be cool if we visually show which "state" is actually being used (i.e. written to the `ModsConfig.xml` and actively being used by the game)

**Comment by loveboylion** [comment]
I agree with both of u but , its just that Users sometimes have RimSort running in the background while playing RimWorld and
if by mistake change anything will it autosave mod list in memory or actual disk ? 
if in memory only then what if Users just closes RimSort or Restarts Windows what happens to the mod list ?

I would suggest a logging feature as well with this if possible so that all mod list changes gets logged

and basically https://github.com/RimSort/RimSort/issues/86 will not be required and can be closed ?

**Comment by jefferyharrell** [comment]
RimWorld only reads `ModsConfig.xml` at game startup time. You can change it all you want while the game is running, and the changes will only take effect next time the game is started. So that's not a problem. If you want to launch RimWorld, leave it running, then mess around with your mod list that's totally fine. RimWorld won't even notice.

**Comment by loveboylion** [comment]
i rephrased and edited the previous question to be more clear

**Comment by jefferyharrell** [comment]
> 1. if by mistake change anything will it autosave mod list in memory or actual disk ?

On disk.

> 3. I would suggest a logging feature as well with this if possible so that all mod list changes gets logged

Same answer as before: I think it's a decent idea, but I can't right this second think how to do it. Needs more talking.

> 4. basically [Implement end-user notification if mod list changes detected and the mod list was not saved #86](https://github.com/RimSort/RimSort/issues/86) will not be required and can be closed ?

Sounds like it, but that's just my two cents.
